File: backend/src/main/java/com/chatbotgen/aplicacao/model/Subscription.java
package com.chatbotgen.aplicacao.model;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.Index;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.Table;
import jakarta.validation.constraints.NotNull;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import java.time.LocalDateTime;

/**
 * Entidade que representa uma assinatura de plano por uma organização.
 * 
 * @author Sistema
 */
@Entity
@Table(name = "subscriptions", indexes = {
    @Index(name = "idx_subscriptions_organization_id", columnList = "organization_id"),
    @Index(name = "idx_subscriptions_plan_id", columnList = "plan_id"),
    @Index(name = "idx_subscriptions_status", columnList = "status")
})
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Subscription {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne
    @JoinColumn(name = "organization_id", nullable = false)
    @NotNull(message = "Organização é obrigatória")
    private Organization organization;

    @ManyToOne
    @JoinColumn(name = "plan_id", nullable = false)
    @NotNull(message = "Plano é obrigatório")
    private Plan plan;

    @Column(nullable = false, length = 20)
    @Enumerated(EnumType.STRING)
    @NotNull(message = "Status é obrigatório")
    @Builder.Default
    private SubscriptionStatus status = SubscriptionStatus.ACTIVE;

    @Column(nullable = false, name = "start_date")
    @NotNull(message = "Data de início é obrigatória")
    private LocalDateTime startDate;

    @Column(name = "end_date")
    private LocalDateTime endDate;

    @CreationTimestamp
    @Column(nullable = false, updatable = false, name = "created_at")
    private LocalDateTime createdAt;

    @UpdateTimestamp
    @Column(nullable = false, name = "updated_at")
    private LocalDateTime updatedAt;

    /**
     * Verifica se a assinatura está ativa no momento.
     * 
     * @return true se estiver ativa
     */
    public boolean isActive() {
        if (status != SubscriptionStatus.ACTIVE) {
            return false;
        }
        
        LocalDateTime now = LocalDateTime.now();
        if (startDate != null && now.isBefore(startDate)) {
            return false;
        }
        
        if (endDate != null && now.isAfter(endDate)) {
            return false;
        }
        
        return true;
    }
}

File: frontend/types/subscription.ts
import type { Organization } from './organization'
import type { Plan } from './plan'

/**
 * Status de uma assinatura
 */
export enum SubscriptionStatus {
  ACTIVE = 'ACTIVE',
  INACTIVE = 'INACTIVE',
  EXPIRED = 'EXPIRED',
  CANCELLED = 'CANCELLED',
}

/**
 * Assinatura de plano por uma organização
 */
export interface Subscription {
  id: number
  organization: Organization
  plan: Plan
  status: SubscriptionStatus
  startDate: string
  endDate?: string | null
  createdAt: string
  updatedAt: string
}

/**
 * Request para criar assinatura
 */
export interface CreateSubscriptionInput {
  organizationId: number
  planId: number
  startDate: string
  endDate?: string | null
}

/**
 * Resposta com limites atuais do plano
 */
export interface PlanLimitsResponse {
  maxUsers: number
  maxChatbots: number
  maxPhoneNumbers: number
  currentUsers: number
  currentChatbots: number
  currentPhoneNumbers: number
}

